colcon_core/dependency_descriptor.py

```diff
@@ -0,0 +1,36 @@
+# Copyright 2016-2018 Dirk Thomas
+# Licensed under the Apache License, Version 2.0
+
+
+class DependencyDescriptor:
+    """A descriptor for a package dependency."""
+
+    def __init__(self, name, metadata=None):
+        """
+        Descriptor for a package dependency.
+
+        :param str name: Name of the declared dependency
+        :param dict metadata: Any metadata associated with
+        the dependency
+        """
+        self.name = name
+
+        if metadata is None:
+            self.metadata = {}
+        else:
+            self.metadata = metadata
+
+    def __eq__(self, other):  # noqa: D105
+        return self.name == other.name
+
+    def __lt__(self, other):  # noqa: D105
+        return self.name < other.name
+
+    def __hash__(self):  # noqa: D105
+        return hash(self.name)
```

Comment:
Generally this is dangerous since after an instance has been hashed the name could be changed which would result in a different hash. I don't have a better proposal though. Maybe prevent setting the `name` outside of the constructor in this case?

Reply:
Converted name to be a read-only property. 

The other issue I see here is that two DependencyDescriptor with different metadata but the same name will be considered equal. Should __eq__ contain the metadata dictionary too? That would require everything stored in metadata to be comparable, is that okay?

Reply:
Another option would be to just not override `__eq__` and wait until there is a use case for it.

Reply:
Overriding `__eq__` is nice for unit testing, not sure if that is enough of a reason? 

Should I just make `__eq__` use `(self.name, self.metadata)`, and have `__hash__` just use `self.name`?

Reply:
It's needed in a couple different tests where we check dependencies. It would also be needed by any consumers of colcon-core that want to test using DependencyDescriptor.

I think keeping `__eq__` would be simpler.

Reply:
> It would also be needed by any consumers of colcon-core that want to test using DependencyDescriptor.

Do we know what potential consumers want to test for? If not, I would rather leave it out for now until that is the case. We can always add it later when there is a concrete need for it.

Reply:
Looking into what happens when I remove `__eq__`, we get duplicates in the set. If I remove `__eq__` then I see unit tests where the same dependency is listed twice in the set.

    def test_get_dependencies():
        d1 = PackageDescriptor('/some/path')
        d1.name = 'self'
        d1.dependencies['build'].add(DependencyDescriptor('build-depend'))
        d1.dependencies['build'].add(DependencyDescriptor('depend'))
        d1.dependencies['run'].add(DependencyDescriptor('run-depend'))
        d1.dependencies['run'].add(DependencyDescriptor('depend'))
    >       assert check_dependencies(d1.get_dependencies(),
                                  ['build-depend', 'run-depend', 'depend'])
    E       AssertionError: assert False
    E        +  where False = check_dependencies({{_name: build-depend, metadata: {}}, {_name:     depend, metadata: {}}, {_name: depend, metadata: {}}, {_name: run-depend, metadata: {}}},     ['build-depend', 'run-depend', 'depend'])
    E        +    where {{_name: build-depend, metadata: {}}, {_name: depend, metadata: {}},     {_name: depend, metadata: {}}, {_name: run-depend, metadata: {}}} = <bound method PackageDescriptor.get_dependencies of <colcon_core.package_descriptor.PackageDescriptor object at 0x1074a3db0>>()
    E        +      where <bound method PackageDescriptor.get_dependencies of <colcon_core.package_descriptor.PackageDescriptor object at 0x1074a3db0>> = <colcon_core.package_descriptor.PackageDescriptor object at 0x1074a3db0>.get_dependencies

Above we can see that `d1.get_dependencies()` returns a set containing the "depend" dependency twice. If I define `__eq__` as `(this.name, this.metadata) == (other.name, other.metadata)` then there are no duplicates. 

It's my understanding that an object with a matching hash is not necessarily equal. So `obj1.hash == obj2.hash` doesn't imply `obj1 == obj2`, but `obj1 == obj2` implies `obj1.hash == obj2.hash`. This means that the set semantics will keep both copies of the object if `__eq__` is not defined.

Reply:
> but `obj1 == obj2` implies `obj1.hash == obj2.hash`.

Indeed, looking at the Python documentation of [__hash__](https://docs.python.org/3/reference/datamodel.html#object.__hash__):

    "The only required property is that objects which compare equal have the same hash value"

> Above we can see that d1.get_dependencies() returns a set containing the "depend" dependency twice.

The return value of `get_dependencies()` *should* contain "duplicates" because two dependencies with the same name do not necessarily have the same version constraints.

So if you store dependencies in a `set` (as this function does in the return value) it *must not* de-duplicate dependencies solely based on their name since the metadata might be different. As a direct consequence the hash can't be computed based on the name alone.

The question if dependencies which have the same name and metadata should be de-duplicated is a separate one: the function itself could (not must) de-duplicate dependencies which are equal in all the properties. That could either happen within the logic of the function or if `__hash__` is defined across all properties (instead of relying on the object id). And again in this case my previous concern applies again: exposing the metadata which allows callers to manipulate the data *after* a dependency has been hashed is dangerous.

Therefore I am still advocating to neither define `__hash__` nor `__eq__` since I don't see the need for them.